File: android-native/app/src/main/java/com/example/kpkn/data/db/DatabaseBackupHelper.kt
package com.example.kpkn.data.db

import android.content.Context
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object DatabaseBackupHelper {

    fun createSnapshot(context: Context): String {
        val db = KpknDatabase.getInstance(context)
        // Checkpoint completo para asegurar que todos los logs WAL se consoliden en el archivo .db principal
        db.openHelper.writableDatabase.query("PRAGMA wal_checkpoint(FULL)").close()

        val dbFile = context.getDatabasePath("kpkn.db")
        if (!dbFile.exists()) {
            throw Exception("Base de datos no encontrada para respaldar")
        }

        val snapshotDir = File(context.filesDir, "snapshots").apply { mkdirs() }
        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        val snapshotFile = File(snapshotDir, "kpkn_snapshot_$timestamp.db")

        copyFile(dbFile, snapshotFile)
        return snapshotFile.name
    }

    fun listSnapshots(context: Context): List<File> {
        val snapshotDir = File(context.filesDir, "snapshots")
        if (!snapshotDir.exists()) return emptyList()
        return snapshotDir.listFiles { _, name -> name.startsWith("kpkn_snapshot_") && name.endsWith(".db") }
            ?.sortedByDescending { it.lastModified() }
            ?: emptyList()
    }

    fun restoreSnapshot(context: Context, snapshotFile: File) {
        if (!snapshotFile.exists()) {
            throw Exception("Archivo de snapshot no encontrado")
        }

        // 1. Cerrar de forma segura la instancia activa de Room
        KpknDatabase.closeInstance()

        val destDbFile = context.getDatabasePath("kpkn.db")

        // 2. Eliminar archivos temporales de log WAL/SHM para evitar corrupción por colisión de logs antiguos
        context.getDatabasePath("kpkn.db-wal").delete()
        context.getDatabasePath("kpkn.db-shm").delete()

        // 3. Reemplazar el archivo de base de datos principal
        copyFile(snapshotFile, destDbFile)
    }

    fun deleteSnapshot(snapshotFile: File): Boolean {
        return if (snapshotFile.exists()) snapshotFile.delete() else false
    }

    private fun copyFile(source: File, dest: File) {
        FileInputStream(source).use { inStream ->
            FileOutputStream(dest).use { outStream ->
                val buffer = ByteArray(1024)
                var bytesRead: Int
                while (inStream.read(buffer).also { bytesRead = it } > 0) {
                    outStream.write(buffer, 0, bytesRead)
                }
            }
        }
    }
}

File: ios-native/KPKNFit/KPKNFit/Data/Db/DatabaseBackupHelper.swift
import Foundation
import SQLite3

internal enum DatabaseBackupHelper {

    static func createSnapshot() throws -> String {
        let db = KpknDatabase.instance()
        let internalDb = db.internalDb
        sqlite3_exec(internalDb.handle, "PRAGMA wal_checkpoint(FULL)", nil, nil, nil)

        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let dbFile = docs.appendingPathComponent("kpkn.db")
        guard FileManager.default.fileExists(atPath: dbFile.path) else {
            throw NSError(domain: "DatabaseBackupHelper", code: 1, userInfo: [NSLocalizedDescriptionKey: "Base de datos no encontrada para respaldar"])
        }

        let snapshotDir = docs.appendingPathComponent("snapshots")
        try FileManager.default.createDirectory(at: snapshotDir, withIntermediateDirectories: true)

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        formatter.locale = Locale.current
        let timestamp = formatter.string(from: Date())
        let snapshotFile = snapshotDir.appendingPathComponent("kpkn_snapshot_\(timestamp).db")

        try copyFile(from: dbFile, to: snapshotFile)
        return snapshotFile.lastPathComponent
    }

    static func listSnapshots() -> [URL] {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let snapshotDir = docs.appendingPathComponent("snapshots")
        guard FileManager.default.fileExists(atPath: snapshotDir.path) else { return [] }
        let files = (try? FileManager.default.contentsOfDirectory(at: snapshotDir, includingPropertiesForKeys: nil)) ?? []
        return files
            .filter { $0.lastPathComponent.hasPrefix("kpkn_snapshot_") && $0.pathExtension == "db" }
            .sorted { $0.lastPathComponent > $1.lastPathComponent }
    }

    static func restoreSnapshot(snapshotFile: URL) throws {
        guard FileManager.default.fileExists(atPath: snapshotFile.path) else {
            throw NSError(domain: "DatabaseBackupHelper", code: 2, userInfo: [NSLocalizedDescriptionKey: "Archivo de snapshot no encontrado"])
        }

        KpknDatabase.closeInstance()

        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let destDbFile = docs.appendingPathComponent("kpkn.db")

        try? FileManager.default.removeItem(at: docs.appendingPathComponent("kpkn.db-wal"))
        try? FileManager.default.removeItem(at: docs.appendingPathComponent("kpkn.db-shm"))

        try copyFile(from: snapshotFile, to: destDbFile)
    }

    static func deleteSnapshot(snapshotFile: URL) -> Bool {
        guard FileManager.default.fileExists(atPath: snapshotFile.path) else { return false }
        return (try? FileManager.default.removeItem(at: snapshotFile)) != nil
    }

    private static func copyFile(from source: URL, to dest: URL) throws {
        let data = try Data(contentsOf: source)
        try data.write(to: dest, options: .atomic)
    }
}
